qgis/QGIS_Tasks.md

```diff
@@ -0,0 +1,454 @@
+        try:
+            for i, bbox in enumerate(self.grid):
+                if self.isCanceled():
+                    return False
+
+                # Get the filepath and create the output directory (if necessary)
+                x, y = (
+                    bbox.boundingBox().xMinimum(),
+                    bbox.boundingBox().yMinimum(),
+                )
+                hemisphere = "S" if y < 0 else "N"
+                dir_name = "%s%02d" % (hemisphere, abs(y))
+                directory = self.output_dir.joinpath(dir_name)
+                directory.mkdir(exist_ok=True)
+                tile_name = "%s%02d%s%03d.hgt" % (
+                    hemisphere,
+                    abs(y),
+                    "W" if x < 0 else "E",
+                    abs(x),
+                )
+                filepath = self.output_dir.joinpath(dir_name, tile_name)
+                self.tiles.append(filepath)
+                if filepath.is_file():
+                    QgsMessageLog.logMessage(
+                        message=f"Already downloaded tile {tile_name}",
+                        level=Qgis.Info,
+                    )
+                    continue
+
+                # get the new tile if it didn't exist
+                url = QUrl(
+                    f"http://s3.amazonaws.com/elevation-tiles-prod/skadi/{dir_name}/{tile_name}.gz"
+                )
+                request = QNetworkRequest(url)
+                reply: QgsNetworkReplyContent = nam.blockingGet(request)
+                # turn the reply into a file object and write the decompressed file to disk
+                with gzip.GzipFile(
+                    fileobj=BytesIO(reply.content()), mode="rb"
+                ) as gz:
+                    with open(filepath, "wb") as f:
+                        f.write(gz.read())
+
+                self.setProgress((i / len(self.grid)) * 100)
+
+            return True
+        except Exception as e:
+            self.exception = e
+            return False
+```
+The `run()` method of a `QgsTask` always needs to return a boolean value, which is used in `finished()` to evaluate whether the task ran successfully. After initializing the network access manager, which is used for handling the http requests, we wrap the remainder of the method inside a `try`/`except` block to prevent exceptions from being raised. If an exception is caught, we save it and return `False`. In order to give the user control over cancelling the download, we check for this after each iteration by calling `self.isCanceled()`, and return False if the user has canceled the task.
+
+The logic of the actual heavy lifting inside `run()` works as follows: for every box in the grid, we get the southwesternmost coordinates, from wich the directory and tile names are created, as well as the URL to download the tile from. If the file does not exist yet, we download it, extract the `.gz`file and save the result into the respective directory.
```

Comment:
```suggestion
The logic of the actual heavy lifting inside `run()` works as follows: for every box in the elevation tile grid, we get the southwestern-most coordinates (their base lat/lon), from which the directory and tile names are created, as well as the URL to download the tile from. If the file does not exist yet, we download it, extract the `.gz` file in-memory and save the result into the respective directory.

The `run()` method of a `QgsTask` always needs to return a boolean value, which is used in `finished()` to evaluate whether the task ran successfully. After initializing the `QgsNetworkAccessManager`, which is used for handling the HTTP requests, we wrap the remainder of the method inside a `try`/`except` block to prevent exceptions from being raised. If an exception is caught, we save it and return `False`. In order to give the user control over cancelling the download, we check `self.isCanceled()` after each iteration, and return `False` if the user has canceled the task.


```